libs/agno/agno/agent/agent.py

```diff
@@ -4032,25 +4028,42 @@ def read_from_storage(
         Returns:
             Optional[AgentSession]: The loaded AgentSession or None if not found.
         """
-        if self.storage is not None:
+        from time import time
+
+        if self.memory is not None and self.memory.db is not None:
+            log_info(f"Reading AgentSession: {session_id}")
             # Get a single session from storage
-            self.agent_session = cast(AgentSession, self.storage.read(session_id=session_id))
+            self.agent_session = cast(AgentSession, self.memory.read_session(session_id=session_id))
             if self.agent_session is not None:
                 # Load the agent session
                 self.load_agent_session(session=self.agent_session)
-        return self.agent_session
-
+        if self.agent_session is not None:
+            return self.agent_session
+        else:
+            self.team_session_id = cast(str, self.team_session_id)
+            self.agent_id = cast(str, self.agent_id)
+            self.agent_session = AgentSession(
+                session_id=session_id,
+                agent_id=self.agent_id,
+                user_id=user_id,
+                team_session_id=self.team_session_id,
+                runs=None,
+                summary=None,
+                agent_data=self.get_agent_data(),
+                session_data=self.get_session_data(),
+                extra_data=self.extra_data,
+                created_at=int(time()),
+            )
+        
     def write_to_storage(self, session_id: str, user_id: Optional[str] = None) -> Optional[AgentSession]:
         """Save the AgentSession to storage
 
         Returns:
             Optional[AgentSession]: The saved AgentSession or None if not saved.
         """
-        if self.storage is not None:
-            self.agent_session = cast(
-                AgentSession,
-                self.storage.upsert(session=self.get_agent_session(session_id=session_id, user_id=user_id)),
-            )
+        if self.memory.db is not None:
+            self.memory.db.upsert_agent_session(session=self.agent_session)
```

Comment:
better python would be to add a `upsert_agent_session` passthrough method on the memory class itself.